File: Connection/MessagePacket.cs
using System;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LANchat.Connection
{
    public class MessagePacket : AbstractTcpPacket
    {

        [Serializable]
        public class MessageTimePair
        {
            public string Message;
            public long Time;

            [JsonIgnore]
            public static DateTime Since = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);

            public MessageTimePair(string message, long time)
            {
                Message = message;
                Time = time;
            }
        }

        public MessageTimePair MessageAndTime;

        public MessagePacket(string message) : base(PacketDataType.MESSAGE)
        {
            long secondsFromEpoch = (long)(DateTime.Now.ToUniversalTime() - MessageTimePair.Since).TotalSeconds;
            MessageAndTime = new MessageTimePair(message, secondsFromEpoch);
        }

        public MessagePacket(byte[] bytes) : base(PacketDataType.MESSAGE)
        {
            MessageAndTime = JsonSerializer.Deserialize<MessageTimePair>(bytes.Skip(1).ToArray(),
                new JsonSerializerOptions() { IncludeFields = true });
        }

        public override byte[] ToByteArray()
        {
            return Encoding.UTF8.GetBytes(JsonSerializer.Serialize(MessageAndTime, new JsonSerializerOptions()
            { IncludeFields = true })).Prepend(base.ToByteArray()[0]).ToArray();
        }

        public override void Apply(MainWindow mw, ChatClient client)
        {
            mw.ProcessMessage(MessageAndTime.Message, MessageAndTime.Time, client);
        }
    }
}

File: connection/messagepacket.cpp
#include "messagepacket.h"
#include <QDateTime>
#include <QJsonObject>
#include <QJsonDocument>
#include "client.h"

namespace Connection
{
    MessagePacket::MessagePacket(QString message)
        : AbstractTcpPacket(PacketDataType::MESSAGE),
          MessageAndTime(message, QDateTime::currentSecsSinceEpoch())
    { }

    MessagePacket::MessagePacket(QByteArray& bytes)
        : AbstractTcpPacket(PacketDataType::MESSAGE)
    {
        QJsonDocument doc = QJsonDocument::fromJson(bytes);
        MessageAndTime.Message = doc.object().value("Message").toString();
        MessageAndTime.Time = doc.object().value("Time").toInt();
    }

    void MessagePacket::Apply(MainWindow& mw, ChatClient& client)
    {
        mw.ProcessMessage(MessageAndTime, &client);
    }

    QByteArray MessagePacket::ToByteArray()
    {
        QJsonDocument doc(QJsonObject{
            { "Message", MessageAndTime.Message },
            { "Time", QJsonValue((qint64)MessageAndTime.Time) }
        });
        return AbstractTcpPacket::ToByteArray().append(doc.toJson());
    }
} // namespace Connection
